Update `impl`-image, tests and enable tool version strings

se-bi opened the pull request:
Thanks, @eine for the discussion on #1 and adding the _impl_ image.  
(Surprisingly, the documentation was ahead ... :relieved: )

Some ideas came up on that:
- Separating impl images regarding ice40 and ecp5 architectures
- Adding tools of icestorm and prjtrellis, respectively
- Clone Github Projects instead of Tar-archive via codeload, to have a meaningful `--version` return of the builds
- Add some test calls of tools (`--version` or similar)

(see the [CI runs](https://github.com/se-bi/hdl-containers/actions?query=+branch%3Atest%2Fbuild-version++), which are not running unchanged parts and pushing into different DockerHub Repositories
https://github.com/se-bi/hdl-containers/compare/develop...se-bi:test/build-version)

eine commented on impl.dockerfile:
```diff
@@ -1,17 +1,24 @@
+FROM base AS ice40
+
+COPY --from=hdlc/pkg:nextpnr-ice40 /nextpnr /
+COPY --from=hdlc/pkg:icestorm /iceprog /
```
I think that iceprog does not belong here. The purpose of `impl` images is obtaining a bitstream. That is, the same output you'd get from `nextpnr` images, without requiring separated image before (for GHDL, Yosys, etc.). That alone is useful for CI on GitHub Actions, where programming boards is not possible. Furthermore, on Windows and macOS it's not straightforward to program boards from containers. It's easier to use native tools for that.

If you really want an image that includes the content of `impl` plus `prog`, that needs to be another layer. We need to find a name for that.

se-bi replied:
Also agree, iceprog is out as well.

eine commented on nextpnr.dockerfile:
```diff
@@ -12,10 +12,9 @@ RUN apt-get update -qq \
- && cd /tmp/nextpnr \
- && curl -fsSL https://codeload.github.com/YosysHQ/nextpnr/tar.gz/master | tar xzf - --strip-components=1 \
- && cd build \
+RUN git clone --depth 1 https://github.com/YosysHQ/nextpnr.git /tmp/nextpnr \
```
With `--depth 1` tags are not retrieved, are they? Some projects show the version as a difference from the last tag, plug the SHA of the current commit. See, for example `GHDL 1.0-dev (v0.37.0-1123-g8ed35277) [Dunoon edition]`. "Shallow clones" don't work in those cases.

eine commented on nextpnr.dockerfile:
```diff
@@ -29,10 +28,9 @@ RUN mkdir -p /tmp/nextpnr/build \
- && cd /tmp/nextpnr \
- && curl -fsSL https://codeload.github.com/YosysHQ/nextpnr/tar.gz/master | tar xzf - --strip-components=1 \
- && cd build \
+RUN git clone --depth 1 https://github.com/YosysHQ/nextpnr.git /tmp/nextpnr \
```
There is an experimental feature that allows using resources from previous stages without increasing the size of the current stage. See:

- https://github.com/ghdl/docker/blob/master/vunit.dockerfile#L35
- https://github.com/ghdl/docker/blob/master/vunit.dockerfile#L1

That would allow you to clone the repo once only.

Anyway, this is not critical.

se-bi replied:
What about splitting the stages here, and base each build on that same layer, where the git is fetched once?

eine replied:
Yes, that makes sense. Cloning the repo twice was an overlook!

At the same time, `build-all` should be `build-generic`. There is no need for building `nextpnr-ice40` and `nextpnr-ecp5` twice! See https://github.com/hdl/containers/pull/10#discussion_r557044028.

However, if you want to contribute that, better do it in a PR for that only. That will allow us to merge it faster, and #10 might be rebased on top of it.

se-bi replied:
This:
> At the same time, build-all should be build-generic
and
> See [#10 (comment)](https://github.com/hdl/containers/pull/10#discussion_r557044028).

resolves some confusion with building some parts of `nextpnr` twice
and I will integrate that.

eine reviewed (changes requested):
Hi @se-bi! Sorry for the delay in getting back to you.

> - Separating impl images regarding ice40 and ecp5 architectures
> - Adding tools of icestorm and prjtrellis, respectively

Agree to both changes, but not in the same images. That is:

- `impl:ice40`: GHDL + Yosys + nextpnr-ice40
- `impl:ecp5`: GHDL + Yosys + nextpnr-ecp5
- `impl:pnr`: GHDL + Yosys + nextpnr-ice40 + nextpnr-ecp5
- `impl:icestorm`: `impl:ice40` + icestorm
- `impl:prjtrellis`: `impl:ecp5` + prjtrellis
- `impl:latest`: `impl:pnr` + icestorm + prjtrellis

Feel free to implement some of them only (the ones you are most interested in). Also, feel free to discuss the naming.

> Clone Github Projects instead of Tar-archive via codeload, to have a meaningful `--version` return of the builds

While it might be acceptable in build images/stages, it's a no-go for the "production" images. Be careful with that.

> Add some test calls of tools (`--version` or similar)

This is nice! However, checks such as `--version` or `which` should be better contributed to [hdl/smoke-tests](https://github.com/hdl/smoke-tests). smoke-tests is a submodule of this repo, located in `test/smoke-tests`. Test scripts in this repo should first call the relevant smoke-tests and then run further checks.

eine commented on .github/workflows/impl.yml:
```diff
@@ -25,13 +25,20 @@ jobs:
+        docker build -t hdlc/impl:ecp5 --target=ecp5 - < impl.dockerfile
 
-    - run: docker run --rm -v $(pwd):/wrk hdlc/impl /wrk/test/impl.sh
+    - run: docker run --rm -v $(pwd):/wrk hdlc/impl:ecp5 /wrk/test/impl--ecp5.sh
```
This is no longer relevant, as that syntax was removed.

eine commented on .github/workflows/impl.yml:
```diff
@@ -25,13 +25,20 @@ jobs:
+      #with:
+      #  args: ./test/impl.sh
+
+    - run: docker run --rm -v $(pwd):/wrk hdlc/impl:ice40 /wrk/test/impl--ice40.sh
```
Not relevant anymore.

eine commented:
@se-bi, I suggest squashing your commits in a single one, and then rebasing that on top of master. I believe that will make it easier to deal with possible conflicts. You can later split them again, if you want to keep several commits in the history. For instance, the change related to adding `git` in `hdlc/build:build` might be handled as a separated PR. That would make this one easier to handle.

se-bi commented on ghdl-yosys-plugin.dockerfile:
```diff
@@ -2,6 +2,7 @@ FROM hdlc/yosys AS base
 
 RUN apt-get update -qq \
  && DEBIAN_FRONTEND=noninteractive apt-get -y install --no-install-recommends \
+    git \
```
> Anyway, I think that ghdl-yosys-plugin does not use git nor provide any version info.
Indeed, took this out.

se-bi commented:
Hey @eine, sorry for the delay.

> @se-bi, I suggest squashing your commits in a single one, and then rebasing that on top of master. I believe that will make it easier to deal with possible conflicts. You can later split them again, if you want to keep several commits in the history. 

I rebased everything and squashed some commits.

> For instance, the change related to adding git in hdlc/build:build might be handled as a separated PR. That would make this one easier to handle.

Should we also make the changes within `nextpnr` which are the base for some in `impl` also in a separate PR?
Since the `dockerBuild` of `impl` would fail without the necessary images and no dependencies of the workflows.

(I tried integrating that dependency once, but didn't work as soon as expected and was not patient enough...)

Btw. thanks @umarcor for the scripts in `.github/bin`, they ease life with a fork :see_no_evil:

eine commented:
> Hey @eine, sorry for the delay.

No worries at all! The conception of organisation HDL is that all of us are busy and we are already contributing/working on other several open source projects. These repositories are not a priority and we do not need to rush for having features added. The purpose is to provide a robust and reasonably stable set of solutions for the community, which allows us to eventually reduce the maintenance burden in the other repos. That is, if we can provide a container that multiple projects can use, they will use it. If we don't, they won't do CI or they will use their custom solution, which most of the projects are already doing. In other words, this org is a low-pass filter. Sit down, relax, and have fun building these awesome tooling :wink:

> Should we also make the changes within `nextpnr` which are the base for some in `impl` also in a separate PR?
> Since the `dockerBuild` of `impl` would fail without the necessary images and no dependencies of the workflows.

Yes please. We can split this into as many PRs as you wish. We can keep this one as a reference and handle each relevant modification at a time. As said, no rush. Let's find our rythm. Once you/we get familiar with this codebase, the dance will flow.

> (I tried integrating that dependency once, but didn't work as soon as expected and was not patient enough...)

In fact, you cannot make it work. Workflows are independent, thus, executed on independent virtual machines (probably located in different datacenters depending on the day/time). As a result, the images you build in one workflow are not visible to others. Successfully built and tested images need to be pushed to a registry (we use dockerhub), so that later executions of other workflows can use them by name. Yet, you are not allowed to push `hdlc/*` images to the registry. Therefore, you would need to rename that to something different (a namespace you own). That is not supported currently, as that name is hardcoded in all the dockerfiles and workflows. Therefore, I suggest to avoid it because it's a lot of manual effort which cannot be upstreamed here. Instead:

- You can build images locally. Write a shell script with the sequence of `dockerBuild` calls in the correct order.
- You can add a temporary workflow where you copy the steps from other several workflows. E.g., copy the esteps of workflow nextpnr, along with the ones for impl. The workflow itself won't be upstreamed, but when you push it to your fork, all the images will exist in the same job (virtual machine).

Nonetheless, if you want to actually fork this repo and have your own namespace in an image registry, you are entitled to do so; this is free software. We'd prefer if enhancements were contributed here (as we are doing in these PRs), but any user is implicitly invited to build custom variants.

eine reviewed (changes requested):
It looks nice overall! Most of the test scripts, dockerfiles and workflows are good. Just a few fixes and rebasing/splitting.

se-bi commented on .github/workflows/impl.yml:
```diff
@@ -2,6 +2,7 @@
 #   Unai Martinez-Corral
 #
 # Copyright 2019-2021 Unai Martinez-Corral <unai.martinezcorral@ehu.eus>
+# Copyright 2021 Sebastian Birke <git@se-bi.de>
```
Alright,
should I add somebody else, who appears in the git history,
but is not listed there, like yourself @eine here an there?

eine replied:
No need to revisit the history, just add yourself to the list of authors of the files you modify/create (if you want), and copy the remainder of the header as is. Each user/contributor is allowed not to provide a "real" name and not to be added to the list of authors explicitly. That's why 'eine' and others are not shown, and that's also why it's not automated. The git history does already track it from a technical point of view, which is what we need for using git blame and knowing who we can ask about possible future issues.

eine commented:
@se-bi, I think this is all done, isn't it? Can we close it?
